lib/ask/runtime/session.ex

```diff
@@ -673,10 +681,42 @@ defmodule Ask.Runtime.Session do
     end)
   end
 
-  defp falls_in_quota_already_completed?(respondent, flow) do
+  # Must the respondent be rejected?
+  defp must_be_rejected?(
+         respondent,
+         reply_disposition,
+         in_quota_completed_steps,
+         count_partial_results
+       ) do
+    # The respondent disposition is updated by Runtime.Survey after the current step is handled
+    # by Runtime.Session, but the quota bucket could be assigned to the respondent and the quota
+    # incremented before the disposition update happens. So if the disposition should be updated
+    # the respondent also must be tested using their new disposition.
+    respondent_incremented_their_quota? =
+      if Flow.should_update_disposition(respondent.disposition, reply_disposition) do
+        updated_respondent = %{respondent | disposition: reply_disposition}
+
+        Respondent.incremented_their_quota?(respondent, count_partial_results) ||
+          Respondent.incremented_their_quota?(updated_respondent, count_partial_results)
+      else
+        Respondent.incremented_their_quota?(respondent, count_partial_results)
+      end
```

Comment:
Look at this solution and tell what you think 

(I thought it considering that we changed `Repsondent.incrementer_their_quota?` parameters to receive the `bucket_it` and the `disposition`)

```elixir
inc_quota = fn disposition ->  Respondent.incremented_their_quota?(respondent.bucket_id, disposition, count_partial_results) end
    
respondent_inc_quota =
      inc_quota(respondent.disposition) || 
      Flow.should_update_disposition(respondent.disposition, reply_disposition) && inc_quota(reply_disposition)
```
For me is easier to understand, and we avoid duplicating `inc_quota(respondent.disposition)`, but please tell me what **you** think. Also, I haven't run and test it, so it could fail

Reply:
Do you see it better now? And... What do you think about the `Respondent.incremented_their_quota?` function. Was it a good decision? I tried to express in the codebase part of the knowledge we gained in these last days including it in the Respondent module.

Reply:
yes, it looks better now 🙂 
I believe it was a good decision encapsulating that logic and giving it a name. It sure is more descriptive now! 🎉